### PR TITLE
Refactor selection logic for collapsed headings

### LINKED ISSUE #10593: Improve cut/paste, undo, and execute for collapsed cells
> - First, I cut and paste a collapsed group =>  when paste the placeholder telling there are hidden cells is not shown nor are the cells...
> - Second I delete a collapsed group and undo the change => same symptom (title is restored but not the collapsed cells nor the placeholder)
> - Thirdly I click on run on a collapsed group => The selection indicator vanishes and it ran only the title cell. I would expected it to actually run all cells in the collapsed group.
>
> collapsed_cells
>
> From that, it sounds like selecting a collapsed group title is *not* selecting all collapsed cells. I think this should be the preferred implementation to have a more natural behavior. It should also help reducing special code to handle collapsed group.

_Originally posted by @fcollonval in https://github.com/jupyterlab/jupyterlab/issues/10571#issuecomment-876361737_

### LINKED ISSUE #15700: Collapsed cells behave unpredictably when moving with keyboard shortcuts
## Description

When selecting a group of cells collapsed under a markdown heading, it appears that they can be moved as a group, but they cannot. Not only that, very strange things happen depending on whether you are moving up or down:
- the cells can appear to disappear completely
- the cells can appear under a different section entirely.

## Reproduce

1. In a notebook, create a markdown section heading cell.
2. Create some number of cells below the section heading.
3. Collapse the section with the arrow beside the heading.
4. Attempt to move the section around with the keyboard shortcut: control+shift+<arrow key> 
5. This will have different behavior depending on the direction and the surrounding cells, I have made a silent 1 minute video showing the various behavior that I see: https://youtu.be/BgTA_ZnELTY 

## Expected behavior

I expected to be able to collapse a section and move the section around with keyboard shortcuts like I can do with an uncollapsed group of cells. 

## Context

- Operating System and version: Mac OS Sonoma 14.2.1
- Browser and version: Firefox 121.0.1 
- JupyterLab version: Version 4.1.0b2 (testing the beta release, but the behavior appears to be the same for older versions).

### LINKED ISSUE #18139: jupyterlab hangs when dragging codecell into a collapsed part of the notebook
## Description

When having a markdown heading collapsed part of the notebook and you drag a code cell in such a way that it would be inside the collapsed part my whole jupyterlab hangs and becomes completely unresponsive. Reloading the tab does not help, but closing the tab and loading jupyterlab again works.

## Reproduce

1. Start a new notebook
2. Make a markdown cell with heading i.e. `# Heading 1`
3. Put a code cell i.e. `print("Hello there!")`
4. Make another markdown cell with heading i.e. `# Heading 2`
5. Put another code cell after that i.e. `print("This is a test.")`
6. Collapse `Heading 1`
7. Drag and drop the `This is a test.` codecell above `Heading 2`
8. jupyterlab now is completely unresponsive (except scrolling up and down in the notebook) and the notebook looks like this:

Image

## Expected behavior

Code cell vanishes into the collapsed part of the notebook.

## Context
 
- Operating System and version: Archlinux
- Browser and version: Chrome
- JupyterLab version: 4.5.0

### PR DESCRIPTION
**Pull request by Meriem-BenIsmail**
## References

- closes https://github.com/jupyterlab/jupyterlab/issues/10593
- closes https://github.com/jupyterlab/jupyterlab/issues/15700
- closes https://github.com/jupyterlab/jupyterlab/pull/10631 (replaces)
- c-f #18808 

## Code changes

This PR refactors and improves the selection logic for notebook cells under collapsed headings. All selection and deselection entry points (`select`, `deselect`, `deselectAll`, `extendContiguousSelectionTo`, and `activeCellIndex` changes) now consistently select all child cells within a collapsed section. 

This ensures that cut, move, and delete operations apply to the entire collapsed group, preventing partial or inconsistent selection states. 

The changes also add a dedicated helper `_selectCollapsedSection` for collapsed section selection and apply it across all relevant code paths.

## Backwards-incompatible changes
None.

### CODE REVIEW

**Comment by Meriem-BenIsmail** [comment]
### Note 
While working on this, I found that inserting or dragging cells near collapsed sections would consistently crash JupyterLab if the collapsed sections were not temporarily expanded. 
The reason is that hidden child cells of collapsed headers are not present in the `widgets` array, so `WindowedList` operations like insertion or move would go out-of-bounds.
 
To avoid this, we expand collapsed sections when necessary. The `expandedSet` ensures each header is only expanded once per operation, preventing unnecessary UI flicker. This approach preserves correct insertion indices, selection, and active cell behavior.

 If someone finds a way to handle this without expanding collapsed sections, that would be an improvement—but currently, this is the only safe approach I find that avoids crashes.

**Comment by Darshan808** [comment]
Test failures seem related.

**Comment by Meriem-BenIsmail** [comment]
> Test failures seem related.



yes there are snapshots that need to be updated but I wanted to get some feedback about this first because this is the reason for the update. 
> Note
While working on this, I found that inserting or dragging cells near collapsed sections would consistently crash JupyterLab if the collapsed sections were not temporarily expanded.
The reason is that hidden child cells of collapsed headers are not present in the widgets array, so WindowedList operations like insertion or move would go out-of-bounds.
To avoid this, we expand collapsed sections when necessary. The expandedSet ensures each header is only expanded once per operation, preventing unnecessary UI flicker. This approach preserves correct insertion indices, selection, and active cell behavior.
If someone finds a way to handle this without expanding collapsed sections, that would be an improvement—but currently, this is the only safe approach I find that avoids crashes.

It would be great to have some feedback first because this is the only fix that I found. @Darshan808

**Review by brichet — COMMENTED** [review]
Thanks @Meriem-BenIsmail for working on this, looks not easy to handle.

I tested it and still see some unexpected behaviors:

- selecting a collapsed header now select al cells below, which is nice and expected. But it seems that the `multiselect` flag is kept when then selecting a single cell (only one time, then it seems to retrieve the correct behavior)

  [multiselect_background.webm](https://github.com/user-attachments/assets/e5ee5acb-2d8c-4817-a77f-9f08df58a967)

- in the screencast below there are several issues:
  - pasting a cell (or a group of cells) below or above a group of cells expands the last group. Is this expected ?
  - when cutting a group of cells again (immediately after pasting them), the focus on the notebook is lost.
  - sometime cells are lost in the undo operation, but not sure this PR is supposed to fix it.
 
  [data_lost.webm](https://github.com/user-attachments/assets/fcb216fe-6546-4a34-84d3-4253593efb22)

- drag and drop group sometime freeze jupyterlab on my side
  
  [freezing.webm](https://github.com/user-attachments/assets/8952e1d6-0c4a-42e6-b3b3-cbd3e7893920)

File: packages/notebook/src/widget.ts [hunk 1]
```diff
@@ -1944,6 +2006,21 @@ export class Notebook extends StaticNotebook {
     this._activeCell = null;
     super.dispose();
   }
+  getCollapsedParent(cellIndex: number, widgets: Cell[]): MarkdownCell | null {
```
Comment (brichet) [inline packages/notebook/src/widget.ts]: This function doesn't seem to be used.
Is this expected ?

**Comment by Meriem-BenIsmail** [comment]
hey @brichet thanks for trying this PR out !
> pasting a cell (or a group of cells) below or above a group of cells expands the last group. Is this expected ?

Yes, this was intentional. The expansion is currently the only way I found to prevent the crashes that happen when pasting or dragging around collapsed sections. As I mentioned earlier, I'm not sure yet if there is a cleaner alternative — I tried to follow the behavior of other editors like VS Code, which also expand sections when structural edits happen nearby.

> drag and drop group sometime freeze jupyterlab on my side

I hadn’t encountered this case yet, so thank you for pointing it out. This area has been quite tricky to work on because the collapsed-section logic is complex, and I haven’t managed to cover all edge cases well. I think I still need a better understanding of how the underlying heading/child relationships are handled internally.

I’ll move this PR back to draft state for now.

**Comment by Meriem-BenIsmail** [comment]
I updated this PR to only handle issues related to inserting or dragging **_cells_** near collapsed sections.

File: packages/notebook/src/widget.ts [hunk 2]
```diff
@@ -646,21 +699,15 @@ export class StaticNotebook extends WindowedList<NotebookViewModel> {
           args.oldIndex,
           -1 * args.oldValues.length
         );
-        // Add default cell if there are no cells remaining.
         if (!sender.length) {
           const model = this.model;
-          // Add the cell in a new context to avoid triggering another
```
Comment (krassowski) [inline packages/notebook/src/widget.ts:657]: Can we keep the comments around here?

**Review by krassowski — COMMENTED** [review]
Thank you for fixing this! Reviewing this change feels daunting as it is a brittle code path which requires careful state management. Would you consider adding a test that would fail before and pass after this change? If this can be replicated with a jest unit test it would be ideal, but if it requires galata/playwright test then that's great too.

File: packages/notebook/src/widget.ts (hunk 2 above)
Comment (Meriem-BenIsmail) [reply packages/notebook/src/widget.ts:657]: yeah will remove them. I kept them to remind  myself what I'm really doing each step, because this was too confusing :cry:

Comment (krassowski) [reply packages/notebook/src/widget.ts:657]: Sorry, I meant to restore the comments which would be dropped by this PR. It is ok to add comments on your new code, but this line along a few other lines appears to have been deleted (possibly by accident?)

Comment (Meriem-BenIsmail) [reply packages/notebook/src/widget.ts:657]: oh yeah sorry misunderstood.  I will add them back again.

**Comment by Meriem-BenIsmail** [comment]
I added 2 Galata tests in `galata/test/jupyterlab/collapsible-headings.test.ts` under `Collapsible Headings; keyboard navigation` that demonstrate the bug and verify the fix:

- Inserting a cell below a collapsed section.
- Moving a cell below collapsed section.

Without this PR: both tests time out, JupyterLab enters an infinite reconciliation loop between the widget list and the cell model, freezing the tab entirely.

With this PR: both tests pass, the collapsed section is expanded before the insertion/move, keeping the widget list and cell model in sync, and the cell lands at the correct position.

I'm not sure this PR fixes https://github.com/jupyterlab/jupyterlab/issues/10593. I tried it a few times, it seems to be solved but maybe it needs to be tested by more people @krassowski

**Review by krassowski — COMMENTED** [review]
This looks good to me, though there is a lot of places where off by one error could creep in due to index handling - I will ask copilot to review.

File: packages/notebook/src/widget.ts [hunk 3]
```diff
@@ -2006,24 +2094,42 @@ export class Notebook extends StaticNotebook {
...
+      if (
+        maybeHeader instanceof MarkdownCell &&
+        maybeHeader.headingCollapsed &&
+        maybeHeader.numberChildNodes > 0 &&
+        i + maybeHeader.numberChildNodes >= index &&
+        Private.selectedProperty.get(maybeHeader)
+      ) {
+        return true;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (krassowski) [reply packages/notebook/src/widget.ts]: I actually prefer the logic to be simpler first rather than fast at first. I am ok if you want to implement this suggestion but this is not something I would hold this PR on as it fixes a serve problem with UI.

File: galata/test/jupyterlab/collapsible-headings.test.ts [hunk 4]
```diff
@@ -298,4 +298,38 @@ test.describe('Collapsible Headings; keyboard navigation', () => {
...
+    await page.keyboard.press('ArrowDown');
+    await page.keyboard.press('ArrowDown');
+
+    await page.keyboard.press('Control+Shift+ArrowUp');
+    await page.waitForTimeout(200);
+
+    expect(
+      await (await page.notebook.getNotebookInPanelLocator())!.screenshot()
+    ).toMatchSnapshot('move_cell_below_collapsed_section.png');
+  });
```
[the remark this answers is not in the document]
Comment (krassowski) [reply galata/test/jupyterlab/collapsible-headings.test.ts:336]: I think that's fine, the code path are qurivalent.

File: packages/notebook/src/widget.ts [hunk 5]
```diff
@@ -1861,6 +1914,17 @@ export class Notebook extends StaticNotebook {
...
 
     this._ensureFocus();
+    if (
+      cell instanceof MarkdownCell &&
+      cell.numberChildNodes > 0 &&
+      cell.headingCollapsed
+    ) {
+      for (let i = newValue; i <= newValue + cell.numberChildNodes; i++) {
+        if (this.widgets[i]) {
+          Private.selectedProperty.set(this.widgets[i], true);
```
[the remark this answers is not in the document]
Comment (krassowski) [reply packages/notebook/src/widget.ts]: I guess this could be fine. @Meriem-BenIsmail if this is intentional, can you add a comment explaining it? If not, I do not have a strong preference on how to address it (reuse `select()` or jut emit `_selectionChanged` from here).

File: packages/notebook/src/widget.ts [hunk 6]
```diff
@@ -1941,32 +2005,56 @@ export class Notebook extends StaticNotebook {
...
+      }
+
+      this.deselectAll();
+      originallySelected.forEach((selected, i) => {
         if (selected) {
-          this.select(this.widgets[to + idx]);
+          const newIndex =
+            targetIndex > from ? targetIndex - n + i + 1 : targetIndex + i;
+          this.select(this.widgets[newIndex]);
         }
```
[the remark this answers is not in the document]
Comment (krassowski) [reply packages/notebook/src/widget.ts]: Testing on binder I did not get a crash initially but with some combination of notebook-level undo (Z) and redo (shift z) I did end up in a hanged state eventually.

[the remark this answers is not in the document]
Comment (krassowski) [reply packages/notebook/src/widget.ts]: Which might be unrelated to the suggestion above, but does highlight that there are some edge cases that we still need to handle.

[the remark this answers is not in the document]
Comment (Meriem-BenIsmail) [reply packages/notebook/src/widget.ts]: yeah, the suggestion seems valid to guard against out-of-range indices, I'll apply it.

> Testing on binder I did not get a crash initially but with some combination of notebook-level undo (Z) and redo (shift z) I did end up in a hanged state eventually.

I couldn't reproduce the crash myself but yeah it still needs a lot of work and testing I guess. So might be best to keep this issue open https://github.com/jupyterlab/jupyterlab/issues/10593.

**Comment by krassowski** [comment]
It looks like I can reproduce a crash quite reliably:

https://github.com/user-attachments/assets/d4201cc4-7110-413c-9953-a5ade01d18fc

**Comment by krassowski** [comment]
This is the infinite loop:

image

image

This looks fundamentally wrong

File: packages/notebook/src/widget.ts [hunk 7]
```diff
@@ -1861,6 +1914,17 @@ export class Notebook extends StaticNotebook {
...
 
     this._ensureFocus();
+    if (
+      cell instanceof MarkdownCell &&
+      cell.numberChildNodes > 0 &&
+      cell.headingCollapsed
+    ) {
+      for (let i = newValue; i <= newValue + cell.numberChildNodes; i++) {
+        if (this.widgets[i]) {
+          this.select(this.widgets[i]);
```
Comment (seankcn) [inline packages/notebook/src/widget.ts:1924]: As far as I can tell the [issue here](https://github.com/jupyterlab/jupyterlab/pull/18084#issuecomment-4162821738) is the same as original loop that happens whenever the active cell ends up inside a collapsed markdown heading, which I think might be related to this line. There's a call to `set activeCellIndex` at the end of the `deselectAll()` function on line 2159, so when a markdown heading gets deselected it then gets reselected here, and if it's during a drag event it doesn't get deselected again before the cells are moved.
This means if you have a collapsed markdown heading selected and you drag another cell, you can end up with non-contiguous selections:
https://github.com/user-attachments/assets/36057094-4692-4548-b5a6-c6b63a135d4c
I think this is enabling a bunch of unexpected behaviour, including being able to place the active cell inside a the collapsed heading within the selection, triggering the loop.

**Comment by syedwasimwb** [comment]
I can still be able to replicate this. 
https://github.com/jupyterlab/jupyterlab/issues/18294#issuecomment-3815477406

which is closed as duplicate. 

Do we have an ETA on the fix?

**Comment by savchenko** [comment]
If that's of any help - here is how I am able to replicate this:

1. Have a markdown (`# Foo`) and code cell directly underneath it.
2. Click on the tiny triangle next to the markdown cell to fold it
3. Drag'n'drop another cell between the markdown and code.

Jupyter Lab will hang indefinitely and peg CPU at 100%:
image

**Comment by MUFFANUJ** [comment]
I think one systemic gap still remains: collapse reconciliation is still split across multiple async entrypoints that can complete out of order.

In the current head:

`refreshCellCollapsed` + `expandParent` are triggered from separate signals in `notebook-extension/src/index.ts` (~L2447-L2500)
`setHeadingCollapse` re-enters via `requestAnimationFrame` when headings are unresolved in `notebook/src/actions.tsx` (~L2191-L2195)
`_onCellCollapsed` applies async completion without stale-completion guarding in `notebook/src/widget.ts` (~L873-L883)
This can still allow older async work to overwrite newer user intent during non-trivial sequences (`move`/`cut`/`paste`/`undo`/`execute`/`toggle`), which matches the “cycle” concern raised in this thread.

I think the fix should converge to:

One serialized reconciliation flow per notebook
Versioned requests, so stale async completions are ignored
Deterministic continuation after heading resolution instead of time-based retry
Regression coverage for complex collapsed-section operations (not only toolbar add above/below).

**Comment by krassowski** [comment]
FYI @afshin there is now also https://github.com/jupyterlab/jupyterlab/pull/18808

**Comment by afshin** [comment]
@krassowski the plan is to move the tests from this PR into that PR and then close this one?

**Comment by krassowski** [comment]
Yes, but only if the new PR turns out to work better than this one, which I do not know for sure yet.

**Comment by krassowski** [comment]
Just to surface it, the plan is to continue with this PR and focus it on user-facing changes (since the crash itself was fixed by https://github.com/jupyterlab/jupyterlab/pull/18808).

File: packages/notebook/src/widget.ts (hunk 7 above)
Comment (afshin) [reply packages/notebook/src/widget.ts]: I *think* moving the `newValue`/`oldValue` up before the loop addresses this, if I understood correctly. So the new test I added failed before moving the guard but passes now.

File: packages/notebook/test/widget.spec.ts [hunk 8]
```diff
@@ -1201,6 +1201,66 @@
...
+          nbformat_minor: 5
+        });
+
+        const heading = widget.widgets[0] as MarkdownCell;
+        await heading.getHeadings();
+        heading.numberChildNodes = 1;
+        heading.headingCollapsed = true;
+
+        // Navigate away then back so active index changes to validate select.
+        widget.activeCellIndex = 1;
```
[the remark this answers is not in the document]
Comment (afshin) [reply packages/notebook/test/widget.spec.ts]: This is necessary in the context of the test.

**Review by afshin — APPROVED** [review]
Thank you @Meriem-BenIsmail! I made two small changes (one about signal emission redundancy, and one moving the guard condition higher up in the setter) and updated tests.

**Comment by krassowski** [comment]
I am not sure if this is a pre-existing issue but testing on binder I see a cells being lost when moving a collapsed section:


https://github.com/user-attachments/assets/59b39b88-1fba-4e3d-bd3f-2b8bcfb6d21e

**Comment by krassowski** [comment]
Probably a better recording:


https://github.com/user-attachments/assets/7d25fbbb-c59d-4773-885b-ae78c98738bf

**Comment by krassowski** [comment]
It seems to be a pre-existing issue I see the same on 4.5.7:


https://github.com/user-attachments/assets/27cf5731-29b5-4fb6-b2a0-8df62bd0c87f

**Comment by krassowski** [comment]
The same happens with copy-paste on this PR:

https://github.com/user-attachments/assets/5ac4bd22-5972-40e6-a516-711e9e1889b3

I suppose I am not requesting changes as I do not have an opinion on the code changes here, I am just highlighting that this does not seem to solve either:
- https://github.com/jupyterlab/jupyterlab/issues/15700 nor
-  https://github.com/jupyterlab/jupyterlab/issues/10593

**Comment by Meriem-BenIsmail** [comment]
@krassowski when you select a collapsed section, the cells inside it don't seem to be selected as well in the screencasts you shared, otherwise it should be working. 

I don't see the whole section being selected (all the section being blue). 

https://github.com/user-attachments/assets/4c2390d2-578a-4d15-9f32-7263f69f1623

**Comment by krassowski** [comment]
@Meriem-BenIsmail can you try testing on Binder?

**Comment by Meriem-BenIsmail** [comment]
hmm yeah i seem to have the same result as you with Binder but locally It seems to work for me and I don't have any changes other than the ones in this PR.

**Comment by krassowski** [comment]
Looking at your recording, I see a state which I recall earlier and then I think it worked. I wonder if maybe you have a stale state locally compared to what is on the remote branch (maybe the build did not clean previous artifacts or maybe you are at a different commit?).

**Comment by Meriem-BenIsmail** [comment]
**Update:** 
I added some logic in `actions.tsx` to fix the case when pasting below a collapsed markdown heading. Previously, pasted cells appeared directly below the header, placing them inside the collapsed section instead of after it.
Now it correctly inserts after all its children. 

Tested on Binder @krassowski @afshin 

https://github.com/user-attachments/assets/27ea79e6-c9b5-44e1-a5a2-e9ed24d5dfcc

**Comment by krassowski** [comment]
There is still something odd happening with selection here I think:


https://github.com/user-attachments/assets/c1e65b42-40d1-4e3b-8a69-f060f029fb2f

**Comment by Meriem-BenIsmail** [comment]
I tried to recreate your screencast on Binder @krassowski 

https://github.com/user-attachments/assets/6509d419-6e8a-4341-8a8a-1d9e13c20027

File: packages/notebook/src/widget.ts [hunk 9]
```diff
@@ -1937,6 +1937,17 @@ export class Notebook extends StaticNotebook {
...
+        }
+      }
+    }
```
Comment (krassowski) [inline packages/notebook/src/widget.ts]: I am still able to get into discontinuous selection state. I believe this is because of the highlighted addition and that `this.deselectAll()` added does not solve the issue:

Image

Comment (Meriem-BenIsmail) [reply packages/notebook/src/widget.ts]: I removed the unnecessary loop through all hidden cells since we already select them in `_selectCollapsedSection` but I'm not sure this will fix the issue you're describing since I couldn't manage to reproduce it. Is there a way to share how you managed to do it ?

**Comment by krassowski** [comment]
Is it expected that clicking next cell after selecting collapsed header adds it to selection rather instead of selecting only the clicked cell? I am surprised by this and as a user this is not what I would expect.


https://github.com/user-attachments/assets/b7de9bf6-3a3a-4872-9a21-3d0038ebc79d

Thank you for your perseverance here, I understand that this is a difficult one with many edge cases. I wonder if adding more unit test as we go would make it easier to work on.

**Comment by Meriem-BenIsmail** [comment]
the issue was that in `deselectAll()`, we were re-triggering the `activeCellIndex` setter which was causing the re-selection. Although I'm not sure of the solution, some tests seem to be failing. Would be great if you could test it on your side @krassowski @afshin 

Update: we can't really remove that part in `deselectAll()`, trying to deselect the cells before selecting.

**Comment by krassowski** [comment]
Looking at the test failures, there is more work to do here. Do you mind if we bump it to 4.7.0?

**Comment by MUFFANUJ** [comment]
The `deselectAll` action does seem to deselect all the cells still  ( the collapsed section/heading stays selected ), I see the comment above the implementation says deselect children, and the test asserts that `deselectAll() `should leave cells selected ( so this seems intentional ).

https://github.com/user-attachments/assets/8d210f89-b808-4ada-81a7-a42a4045c81b

Not sure if this is the case is also intentional.

https://github.com/user-attachments/assets/ca5d78db-08a5-4d7e-bce1-79807892b170

**Comment by MUFFANUJ** [comment]
Also, toggling the collapse seems to select/deselect the children. Maybe more tests should be added to the implemtation.

https://github.com/user-attachments/assets/ed6bc407-48a0-43c8-b188-c010f3a443cf

**Comment by krassowski** [comment]
> The deselectAll action does seem to deselect all the cells still 

This seems fine - it is a pre-existing behaviour that the active cells gets re-selected I see that in 4.5.x

**Comment by MUFFANUJ** [comment]
> This seems fine - it is a pre-existing behaviour that the active cells gets re-selected I see that in 4.5.x

The action in the command palette is less clear on this. Maybe the intended actions should deselect the heading as well, or maybe we should make the command clearer?

**Comment by afshin** [comment]
I think the action we want here is: “Deselect All” clears explicit selection but does not unset the active cell.

For collapsed headings, actions like copy/cut/delete may treat the active or selected heading as representing its children, but that should be derived for those actions rather than by having `activeCellIndex` mutate selection state.

@MUFFANUJ @krassowski does this sound right to you?

---

@Meriem-BenIsmail  if you also agree and we have consensus, I would suggest we can close out this PR with the following changes:

Could you please adjust the implementation so activating a collapsed heading does not select hidden children directly? 

To validate, please add tests that copy/cut/delete still include collapsed children, and add a test that toggling collapse does not unexpectedly change logical selection.

**Review by afshin — APPROVED** [review]
Thank you Meriem! I think this now matches the last request I made, which seemed liked we'd landed on consensus:

- [x] "please adjust the implementation so activating a collapsed heading does not select hidden children directly"
- [x] "please add tests that copy/cut/delete still include collapsed children, and add a test that toggling collapse does not unexpectedly change logical selection"

**Comment by krassowski** [comment]
Most likely for a follow-up PR because this has too long history already: when pressing Shift + Down the selection first covers all cells in the collapsed range (implicitly) but when pressing Shift + Down second time it selects the first cell only; I am not sure if this is good UX, it feels confusing to me.

https://github.com/user-attachments/assets/30b43869-f346-4954-9692-504f0ee79792

**Review by krassowski — APPROVED** [review]
The code changes look minimal and do not extend the public API, so even if this could be better, I think it is an incremental improvement and worth getting out in pre-releases for users to provide feedback on.
